### DISCUSSION TITLE
Box/Container/Section reusable variants (ThemableBox?)

### DISCUSSION THREAD

**Discussion by bhishp** [chakra-ui/chakra-ui#3046] [Q&A] [answered]
I have some visual 'boxes' that are consistent throughout my app, at the moment these are <Box> components with the same style props copy and pasted. This is a lot of DRY and I was thinking of the best way to eliminate that. I can think of 3 ways for this particular case at the moment but I think **Option 3** might be the most suitable as it allows for the number of variants to grow if needed. Is this the best practice or is there a better way to do this?

```ts
// Option 1
const FloatingSectionAsWrappedComponent<BoxProps> = ({ children, ...props }) => (
  <Box
    p="4"
    mb="4"
    borderRadius="base"
    borderWidth="1px"
    borderColor="gray.300"
    bgColor="white"
    {...props}
  >
    {children}
  </Box>
);

// Option 2
const FloatingSectionFactoried = chakra('div', {
  baseStyle: {
    p: '4',
    mb: '4',
    borderRadius: 'base',
    borderWidth: '1px',
    borderColor: 'gray.300',
    bgColor: 'white',
  }
});

// Option 3
type SectionProps = HTMLChakraProps<'div'> & ThemingProps;
const Section = forwardRef<SectionProps, 'div'>((props, ref) => {
  const styles = useStyleConfig('Section', props);
  const rest = omitThemingProps(props);
  return <chakra.div ref={ref} {...rest} __css={styles} />;
});
// in the Provider
<ChakraProvider theme={extendTheme({
  components: {
    Section: {
      variants: {
        'floating': {
          p: '4',
          mb: '4',
          borderRadius: 'base',
          borderWidth: '1px',
          borderColor: 'gray.300',
          bgColor: 'white',
        }
      },
    }
  }})}>


// usage...
(
  <>
    <FloatingSectionAsWrappedComponent>Content</FloatingSectionAsWrappedComponent>
    <FloatingSectionFactoried>Content</FloatingSectionFactoried>
    <Section variant="floating">Content</Section>
  </>
)
```

In essence the Option 3 "Section" is really a "ThemableBox". The chakra Box does not use any style config from the theme, so we can't pass variants to this at the moment. Is there maybe scope to add a TBox to the chakra lib which simple does the above?

As a side note, should the chakra factory allow variants to be passed also? Is there any specific reason it is not available? 

```ts
// It would be cool to be able to pass variants to the factory then we can provide many variants of this section
const SectionFactoried = chakra('div', {
  variants: {
    floating: {
      p: '4',
      mb: '4',
      borderRadius: 'base',
      borderWidth: '1px',
      borderColor: 'gray.300',
      bgColor: 'white',
    }
  }
});

<SectionFactoried variant="floating" />
```

**Comment by chasinhues** [comment, marked as answer]
IMO [layer styles](https://chakra-ui.com/docs/features/text-and-layer-styles#layer-style) would suit this use case perfectly 🚀 

```jsx
const theme = extendTheme({
  layerStyles: {
    floating: {
      p: '4',
      mb: '4',
      borderRadius: 'base',
      borderWidth: '1px',
      borderColor: 'gray.300',
      bgColor: 'white',
    }
  },
})
```

```jsx
<Box layerStyle="floating">I'm floating!</Box>
```

**Reply by osirvent** [reply]
Very interesting! 

So when use `layerStyle` instead of `variant`?

**Reply by bhishp** [reply]
I didn't even know of this feature yet, perfectly suits my needs, thank you!

**Reply by chasinhues** [reply]
Both `variants` and `layerStyles` aim to improve reusability. The difference in use case I would say is that variants are for specific components whereas layer styles are more generic.

For example the `Alert` component has a background for all of its variants. I've encountered a design where I had to create an Alert that just had an outline. In this case, the reusability I required applied specifically to one component, so variant was the way to go.

Your case is a great example of layer style. Much more generic styles that you want to apply to a container/layer/etc. You could apply the layer style to many of the Chakras components. You might want to create a `card` layer style with a border, box shadow, padding, and border radius. And you might want to be able to apply that layer style to a Box, a Flex, a Grid, a Button, etc. In this case, because the application is more generic, layer style works really well.

**Reply by hugobqd** [reply]
How to define light/dark mode colors in layerStyle?

**Reply by chasinhues** [reply]
AFAIK, this isn't possible in layer styles at the moment - although there has been some discussion around it in this issue (among other places) https://github.com/chakra-ui/chakra-ui/issues/2231

The issue linked mentions some alternative workarounds though, might have to go with one of those for the time being until the feature is implemented unfortunately :/
